src/render.rs

```diff
@@ -1,145 +1,343 @@
-//! Render message to a byte buffer
-
 use midi_types::{status::*, MidiMessage};
 
-/// Trait for rendering a MidiMessage into a byte slice.
-pub trait MidiRenderSlice: Sized {
-    /// Render to a slice.
-    ///
-    /// # Panics
+/// This trait abstracts the transport mechanism for the MidiRenderer. An instance of a type that implements this trait can be used by the MidiRenderer to write midi-messages
+pub trait MidiTransport {
+    type Error;
+
+    /// Write a message as series of bytes to the midi transport layer
     ///
-    /// Panics if the slice length is less than 3.
-    fn render_slice(&self, buf: &mut [u8]) -> usize;
+    /// For compatibility this should always be used to write one whole midi-message with a maximum of 4 bytes
+    fn write(&mut self, bytes: &[u8]) -> Result<(), Self::Error>;
 }
 
-//helper to render 3 byte messages
-fn chan3byte<T0: Into<u8> + Copy, T1: Into<u8> + Copy, C: Into<u8> + Copy>(
-    buf: &mut [u8],
-    status: u8,
-    chan: &C,
-    d0: &T0,
-    d1: &T1,
-) -> usize {
-    let chan: u8 = (*chan).into();
-    let status = status | chan;
-    for (o, i) in buf.iter_mut().zip(&[status, (*d0).into(), (*d1).into()]) {
-        *o = *i;
-    }
-    3
+/// The MidiRenderer takes MIDI messages and writes them to the underlying transport, the boolean const generic R enables or disables rendering running status for midi messages
```

Comment:
you renamed the generic `R` to `RUNNING_STATUS` but haven't updated the doc string

Reply:
🤦🏼 thanks! I'll fix this!